neuralhydrology/modelzoo/transformer.py

```diff
@@ -0,0 +1,167 @@
+        self.encoder = nn.TransformerEncoder(encoder_layer=encoder_layers, 
+                                             num_layers=cfg.transformer_nlayers,
+                                             norm=None) 
+
+        # head (instead of a decoder)
+        self.dropout = nn.Dropout(p=cfg.output_dropout)
+        self.head = get_head(cfg=cfg, n_in=self.encoder_dim, n_out=self.output_size)
+
+        # init weights and biases 
+        self._reset_parameters()
+
+    def _reset_parameters(self): 
+        initrange = 0.1
+        for layer in self.encoder.layers:
+            layer.linear1.weight.data.uniform_(-initrange, initrange)
+            layer.linear1.bias.data.zero_()
+            layer.linear2.weight.data.uniform_(-initrange, initrange)
+            layer.linear2.bias.data.zero_()
+        self.embedding.weight.data.uniform_(-initrange, initrange)
+        self.embedding.bias.data.zero_()
+
+    def forward(self, data: Dict[str, torch.Tensor], **kwargs) -> Dict[str, torch.Tensor]:
+        """Perform a forward pass on a transformer model without decoder.
+
+        Parameters
+        ----------
+        data : Dict[str, torch.Tensor]
+            Dictionary, containing input features as key-value pairs.
+
+        Returns
+        -------
+        Dict[str, torch.Tensor]
+            Model outputs and intermediate states as a dictionary. 
+                - `y_hat`: model predictions of shape [batch size, sequence length, number of target variables].
+        """
+        # transpose to [seq_length, batch_size, n_features]
+        x_d = data['x_d'].transpose(0, 1)
+
+        # concat all inputs
+        if 'x_s' in data and 'x_one_hot' in data:
+            x_s = data['x_s'].unsqueeze(0).repeat(x_d.shape[0], 1, 1)
+            x_one_hot = data['x_one_hot'].unsqueeze(0).repeat(x_d.shape[0], 1, 1)
+            x_d = torch.cat([x_d, x_s, x_one_hot], dim=-1)
+        elif 'x_s' in data:
+            x_s = data['x_s'].unsqueeze(0).repeat(x_d.shape[0], 1, 1)
+            x_d = torch.cat([x_d, x_s], dim=-1)
+        elif 'x_one_hot' in data:
+            x_one_hot = data['x_one_hot'].unsqueeze(0).repeat(x_d.shape[0], 1, 1)
+            x_d = torch.cat([x_d, x_one_hot], dim=-1)
+        else:
+            pass
+
+        # embedding
+        x_d = self.embedding(x_d) * math.sqrt(self.embedding_dim)        
+        x_d = self.pos_encoder(x_d, self.posistional_encoding_type)
+
+        # mask past values
+        if self.mask is None or self.mask.size(0) != len(x_d):
+            mask = (torch.triu(torch.ones(len(x_d), len(x_d))) == 1).transpose(0, 1)
+            mask = mask.float().masked_fill(mask == 0, float('-inf')).masked_fill(mask == 1, float(0.0))
+            self.mask = mask.to(x_d.device)
+
+        # encoding
+        output = self.encoder(x_d, self.mask)
+
+        # head
+        pred = self.head(self.dropout(output.transpose(0, 1)))
+
+        return pred
+
+class PositionalEncoding(nn.Module):
+
```

Comment:
As mentioned on Skype, this public class does still miss a doc-string and eventually a little explanation what is happening here. I think this is also the standard temporal encoding from a PyTorch tutorial, no? So maybe add that link as well, if anyone wants to have a look.